src/ellswift.rs

```diff
@@ -0,0 +1,661 @@
+        let mut ser = [0u8; 64];
+        from_hex(hex, &mut ser).unwrap();
+        ElligatorSwift::from_array(ser)
+    }
+    /// Computes a shared secret, just like `shared_secret`, but with a custom hash function
+    /// for computing the shared secret. For compatibility with other libraries, you should
+    /// use `shared_secret` instead, which is already compatible with BIP324.
+    /// The hash function takes three arguments: the shared point, and the ElligatorSwift
+    /// encodings of the two parties and returns a 32-byte shared secret.
+    pub fn shared_secret_with_hasher<F>(
+        ellswift_a: ElligatorSwift,
+        ellswift_b: ElligatorSwift,
+        secret_key: SecretKey,
+        party: ElligatorSwiftParty,
+        mut hash_function: F,
+    ) -> ElligatorSwiftSharedSecret
+    where
+        F: FnMut([u8; 32], [u8; 64], [u8; 64]) -> ElligatorSwiftSharedSecret,
+    {
+        let mut shared_secret = [0u8; 32];
+        let hashfp = hash_callback::<F>;
+        unsafe {
+            secp256k1_ellswift_xdh(
+                ffi::secp256k1_context_no_precomp,
+                shared_secret.as_mut_c_ptr(),
+                ellswift_a.0.as_c_ptr(),
+                ellswift_b.0.as_c_ptr(),
+                secret_key.as_c_ptr(),
+                party.into(),
+                Some(hashfp),
+                &mut hash_function as *mut F as *mut c_void,
+            );
+        }
+        ElligatorSwiftSharedSecret(shared_secret)
+    }
+    /// Encodes a public key into an ElligatorSwift encoding
+    fn ellswift_encode(pk: PublicKey) -> ElligatorSwift {
+        let mut ell_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        unsafe {
+            secp256k1_ellswift_encode(
+                ffi::secp256k1_context_no_precomp,
+                ell_out.as_mut_c_ptr(),
+                pk.as_c_ptr(),
+                [0u8; 32].as_ptr(),
+            );
+        }
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ell_out))
+    }
+    /// Decodes an ElligatorSwift encoding into a [ffi::PublicKey]. This function is a low
+    /// level function, and will be used in internal functions.
+    pub(crate) fn ellswift_decode(ell: ElligatorSwift) -> ffi::PublicKey {
+        unsafe {
+            let mut pk = ffi::PublicKey::new();
+            secp256k1_ellswift_decode(
+                ffi::secp256k1_context_no_precomp,
+                pk.as_mut_c_ptr(),
+                ell.0.as_c_ptr(),
+            );
+            pk
+        }
+    }
+}
+/// The result of [ElligatorSwift::shared_secret], which is a shared secret
+/// computed from the x-only ECDH using both parties' public keys (ElligatorSwift encoded) and our own
+/// private key.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ElligatorSwiftSharedSecret([u8; 32]);
+
+impl Deref for ElligatorSwiftSharedSecret {
+    type Target = [u8; 32];
+
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
```

Comment:
In 6cde4c32dad779e69cb2242ef46fe0646e22d806:

I think we should drop this `Deref` impl. As a rule we avoid using `Deref` to make "transparent newtypes" like this. In this library the only existing use of `Deref` is on `SerializedSignature` which is explicitly a variable-length byte array type and is supposed to behave like an owned slice in every way.

A shared secret, on the other hand, is its own thing (though I appreciate it has no invariants) and should require some explicit action to let you get the bytes out of it. If anything, because these bytes are supposed to be secret so we don't want them to leak accidentally.